### DISCUSSION TITLE
Adapter Maintainers: Upgrading to dbt-core 1.5.0

### DISCUSSION THREAD

**Discussion by dataders** [dbt-labs/dbt-core#7213] [Community]
# Overview

This discussion is for communicating to adapter maintainers the scope of work needed to make use of the changes in 1.5.0. If you have questions and concerns, please ask them here for posterity.

For the guide to upgrading an adapter to be compatible with dbt-core=1.4.0, please see #6624.

The latest release for 1.5.0, `dbt-core==1.5.0rc1`, ([PyPI](https://pypi.org/project/dbt-core/1.5.0rc1/) | [Github](https://github.com/dbt-labs/dbt-core/releases/tag/v1.5.0rc1)).
The following releases are forthcoming with the projected dates:

- :white_check_mark: ~~April 13, 2023: release cut (`1.5.0rc1`)~~
- April 27, 2023: final cut (`1.5.0`)
- May 11, 2023: corresponding minor release for verified adapters not maintained by dbt Labs

# Changes

## Example diffs from dbt Labs-owned adapters

Below gives the changes as of the latest beta release (`1.5.0rc1`)

<https://github.com/dbt-labs/dbt-redshift/compare/v1.4.0...1.5.latest>
<https://github.com/dbt-labs/dbt-bigquery/compare/v1.4.0...1.5.latest>
<https://github.com/dbt-labs/dbt-snowflake/compare/v1.4.0...1.5.latest>
<https://github.com/dbt-labs/dbt-spark/compare/v1.4.0...1.5.latest>

### Support for Py 3.11

Do this if you haven't already! Python 3.11 is the default latest version that is installed without specifying a verson. For example `brew install python`.

### Adapter zone tests

The work is finally finished converting our test suite. This means that, going forward, any adapter test that is useful for more than one adapter will be available to  community and partner adapter maintainers. There are many new tests You can find them all in the `tests/adapter/dbt/tests/adapter/` directory of the dbt-core repo
([link](https://github.com/dbt-labs/dbt-core/tree/main/tests/adapter/dbt/tests/adapter)). Quick callout to some new categories of tests:

- caching
- column_types
- constraints
- hooks
- persist_docs
- simple_copy

### Logging restructuring

TBC (to-be-communicated)

### Upgrade test suite version

Be sure to upgrade the version of `dbt-tests-adapter`

```
# for adapter tests from core
# dbt-tests-adapter~=1.5.0rc1
dbt-tests-adapter~=1.5.0 after final release
```

## New features

as an adapter maintainer, there's three options for new features in dbt Cloud:

1. support the change
2. "stub" the feature as in add in code to let users know that the feature is "not supported"
3. do nothing. if an adapter maintainer does nothing, then, if an end user tries to use that feature, the default implementation of dbt-core will be made which will result in SQL being generated and sent to the underlying data platform, which will likely make the platform throw an error.

### Materialized View, Materialized Tests

Bumped to 1.6! Nothing to do for 1.5 :)

### Model Contracts

**How to support:**

1. Implement `data_type_code_to_name` to convert Python connector return types to strings (example: [dbt-snowflake PR](https://github.com/dbt-labs/dbt-snowflake/pull/484))
2. In `create_table_as`, add calls to two macros that assert column spec equivalence (`get_assert_columns_equivalent`) and template out DDL containing column names, types, and constraints (`get_table_columns_and_constraints`) (example: [dbt-snowflake](https://github.com/dbt-labs/dbt-snowflake/blob/main/dbt/include/snowflake/macros/adapters.sql#L25-L29))
3. In `create_view_as`, add a call to just the macro asserting column spec equivalence - `get_assert_columns_equivalent` (example: [dbt-snowflake](https://github.com/dbt-labs/dbt-snowflake/blob/main/dbt/include/snowflake/macros/adapters.sql#L98-L100))
4. Support the `constraints` data structure ([docs](https://docs.getdbt.com/reference/resource-properties/constraints)) by:
    - Defining a class attribute on the `Adapter` class called `CONSTRAINT_SUPPORT` ([Snowflake example](https://github.com/dbt-labs/dbt-snowflake/blob/main/dbt/adapters/snowflake/impl.py#L43-L49)) that maps constraint types to their support/enforcement values,and
     - Implementing `Adapter` methods [`render_column_constraint`](https://github.com/dbt-labs/dbt-core/blob/main/core/dbt/adapters/base/impl.py#L1307) and [`render_model_constraint`](https://github.com/dbt-labs/dbt-core/blob/main/core/dbt/adapters/base/impl.py#L1384) as necessary and  to render out the appropriate text for each constraint type.
        - For example, it looks like Trino supports **only** `NOT NULL` ([docs](https://trino.io/docs/current/sql/create-table.html)), so `dbt-trino` would implement this method and data structure with only `not_null` constraints supported.
6. Add new functional tests and parametrize them by overriding fixtures:
   - Test classes to subclass:
       - [TestTableConstraintsColumnsEqual](https://github.com/dbt-labs/dbt-core/blob/main/tests/adapter/dbt/tests/adapter/constraints/test_constraints.py#L331)
       - [TestViewConstraintsColumnsEqual](https://github.com/dbt-labs/dbt-core/blob/main/tests/adapter/dbt/tests/adapter/constraints/test_constraints.py#L335)
        - [TestIncrementalConstraintsColumnsEqual](https://github.com/dbt-labs/dbt-core/blob/main/tests/adapter/dbt/tests/adapter/constraints/test_constraints.py#L339)
        - [TestTableConstraintsRuntimeDdlEnforcement](https://github.com/dbt-labs/dbt-core/blob/main/tests/adapter/dbt/tests/adapter/constraints/test_constraints.py#L343)
        - [TestTableConstraintsRollback](https://github.com/dbt-labs/dbt-core/blob/main/tests/adapter/dbt/tests/adapter/constraints/test_constraints.py#L347)
        - [TestIncrementalConstraintsRuntimeDdlEnforcement](https://github.com/dbt-labs/dbt-core/blob/main/tests/adapter/dbt/tests/adapter/constraints/test_constraints.py#L351)
        - [TestIncrementalConstraintsRollback](https://github.com/dbt-labs/dbt-core/blob/main/tests/adapter/dbt/tests/adapter/constraints/test_constraints.py#L357)
        - [TestModelConstraintsRuntimeEnforcement](https://github.com/dbt-labs/dbt-core/blob/main/tests/adapter/dbt/tests/adapter/constraints/test_constraints.py#L418)
   - Example: [test_constraints.py in dbt-snowflake](https://github.com/dbt-labs/dbt-snowflake/blob/main/tests/functional/adapter/test_constraints.py#L40)

**How to stub elegantly:**

In both `create_table_as` and `create_view_as` macros, raise an explicit warning if a model is configured with an enforced contract. For example,

```sql
{% macro <adapter>__create_table_as(...) %}
  {% set contract_config = config.get('contract') %}
  {% if contract_config.enforced %}
     {{exceptions.warn("Model contracts cannot be enforced by <adapter>!")}}
  {% endif %}
{% endmacro %}
```

**What if you do nothing:**

Users will see [new documentation about model contracts](https://docs.getdbt.com/docs/collaborate/publish/model-contracts) and expect to be able to use it!

If your adapter uses the default materializations for table, view, and incremental, and the user configures a model to have an enforced contract, they will see an explicit error:

```bash
ERROR: `data_type_code_to_name` is not implemented for this adapter!
```

If, however, your adapter reimplements those materializations, their configuration may do nothing—which is worse! If you have reimplemented those materializations, and your adapter cannot implement support for contracts, we recommend you raise an explicit warning if they configure a model with an enforced contract.

**Comment by damian3031** [comment]
Example diffs are empty

**Reply by dataders** [reply]
sorry @damian3031 I've just updated it -- does it work now?

**Reply by damian3031** [reply]
yes it works now, thanks

**Comment by damian3031** [comment]
Hi @dataders, what about **Materialized View, Materialized Tests** in 1.5.0? Would `materialized_view` materialization be available in dbt 1.5.0? As it is "TBC" here in discussion, but I don't see it added in 1.5.0rc1

**Reply by Fleid** [reply]
My bad, these 2 items were bumped to 1.6! Nothing to do for 1.5 :)

**Comment by dataders** [comment]
using a comment to track progress across all (known) adapters

- [x] https://github.com/databricks/dbt-databricks/issues/314
- [x] https://github.com/starburstdata/dbt-trino/issues/287
- [x] https://github.com/dbt-athena/dbt-athena/issues/244
- [x] https://github.com/jwills/dbt-duckdb/issues/147
- [x] https://github.com/tglunde/dbt-exasol/issues/68
- [x] https://github.com/MaterializeInc/materialize/issues/18843
- [x] https://github.com/aws-samples/dbt-glue/issues/170
- [x] https://github.com/oracle/dbt-oracle/issues/83
- [x] https://github.com/dremio/dbt-dremio/issues/179
- [x] https://github.com/fal-ai/fal/issues/827
- [x] https://github.com/Teradata/dbt-teradata/issues/75
- [x] https://github.com/databendcloud/dbt-databend/issues/2
- [ ] https://github.com/firebolt-db/dbt-firebolt/issues/103
- [ ] https://github.com/markporoshin/dbt-greenplum/issues/19
- [ ] https://github.com/ClickHouse/dbt-clickhouse/issues/149
- [ ] https://github.com/cloudera/dbt-hive/issues/98
- [ ] https://github.com/aurany/dbt-ibmdb2/issues/37
- [ ] https://github.com/cloudera/dbt-impala/issues/153
- [ ] https://github.com/iomete/dbt-iomete/issues/6
- [ ] https://github.com/mindsdb/dbt-mindsdb/issues/26
- [ ] https://github.com/dbeatty10/dbt-mysql/issues/145
- [ ] https://github.com/chop-analytics/dbt-netezza/issues/19
- [ ] https://github.com/rockset/dbt-rockset/issues/29
- [ ] https://github.com/memsql/dbt-singlestore/issues/20
- [ ] https://github.com/splitgraph/sgr/issues/806
- [ ] https://github.com/dbt-msft/dbt-sqlserver/issues/347
- [ ] https://github.com/codeforkjeff/dbt-sqlite/issues/43
- [ ] https://github.com/microsoft/dbt-synapse/issues/165
- [ ] https://github.com/pingcap/dbt-tidb/issues/50
- [ ] https://github.com/vertica/dbt-vertica/issues/93
- [ ] https://github.com/inferlabs/dbt-infer/issues/7
- [ ] https://github.com/getindata/dbt-flink-adapter/issues/63
- [ ] https://github.com/tanyshak/dbt-upsolver/issues/1

**Comment by aosingh** [comment]
Thank you @dataders  I have started to work on v.1.5.0 for dbt-oracle and have a couple of questions.

- What is the difference in between 3 values for Constraint support/enforced ?

```
ENFORCED = "enforced"
NOT_ENFORCED = "not_enforced"
NOT_SUPPORTED = "not_supported"
```

- Why specify datatype when generating contraints ?

```sql
 create  table dbt_test.my_model
      (id int)
  as
  select 1 as id from dual
```
raises
```
  ORA-01773: may not specify column datatypes in this CREATE TABLE
```
Following works correctly

```sql
 create  table dbt_test.my_model
      (id primary key not null)
  as
  select 1 as id from dual
```

**Reply by dataders** [reply]
hey @aosingh -- great question. I recommend you start by reading https://docs.getdbt.com/reference/resource-properties/constraints#platform-specific-support. the below should also help (thanks @dbeatty10!)

**`ENFORCED`: Supported & enforced**

The model won't even build if it violates this constraint

**`NOT_ENFORCED`: Supported but not enforced**

The platform supports specifying this type of constraint, but a model can still build even if it violates it. It's basically a "fake" constraint only for metadata purposes. This is common for modern cloud data warehouses and less common for legacy databases.

**`NOT_SUPPORTED`: Not supported**

You can't specify this type of constraint for this platform

**more background reading**
https://github.com/dbt-labs/dbt-core/issues/7067#issuecomment-1447281379
https://github.com/dbt-labs/docs.getdbt.com/issues/3282
https://github.com/dbt-labs/docs.getdbt.com/pull/3244#issuecomment-1520382505

**Reply by aosingh** [reply]
Got it @dataders. Thank you for the simple explanation @dbeatty10 . Kudos to the dbt Labs team for implementing model contracts in an elegant manner. I needed to override a few macros to generate the "pre-flight" SQL and the adapter method`render_column_constraint`. All the test cases pass after that
Thank you both!

**Comment by nikitasinghvi** [comment]
Thank you @dataders. I would like to use dbt-core  v1.5.0 with dbt-databricks adapter. Any ETA by when that will be available?

**Reply by dataders** [reply]
any day now! I failed to communicate to the team maintaining dbt-databricks about migration off of the legacy test suite, so the extra work delayed the release. I'd follow https://github.com/databricks/dbt-databricks/ for more information

**Reply by nikitasinghvi** [reply]
Thank you @dataders. I will follow the progress.

**Comment by morsapaes** [comment]
While working on the upgrade, I noticed that [this snippet](https://github.com/dbt-labs/dbt-core/blob/3f2cba0dec9bfdb5cc5dcf942db71fc33b448017/core/dbt/parser/schemas.py#L1026) enforces that contracts can only be applied to `view`, `table`, `incremental` materializations. Our adapter (`dbt-materialize`) ships with some custom materializations that should also support the new contract config (`materializedview`, `source`, `sink`). Is there was a way to enable this that doesn't require patching `dbt-core`?

Something else that is confusing is that the returned error uses _constraints_, rather than _contracts_:

```
❯ dbt run
12:15:15  Running with dbt=1.5.0
12:15:15  Encountered an error:
Parsing Error
  Constraint validation failed for: (models/marts/winning_bids.sql)
  Constraints must be defined in a `yml` schema configuration file like `schema.yml`.
  Only table, view, and incremental materializations are supported for constraints, but found 'materializedview'
```

IIUC, contracts are supported for `table`, `view` and `incremental` materializations, but constraints within contracts can only be specified for `table` and `incremental`. @sungchun12

**Reply by dataders** [reply]
**TL;DR**: what you ask cannot be done without patching dbt-core.

here's my paraphrase of @MichelleArk's response from our [internal slack thread
](https://dbt-labs.slack.com/archives/C034EV1RH4Z/p1683907332835359)

> This is a really good point - we hadn't considered materializations beyond core's table, view, incremental, and ephemeral yet.
> Most straightforward thing we could do is update the check in core to ensure the node's materialization is an unsupported materialization type instead of one of view, table, incremental. It'd be a little more involved to generalize & make this configurable in the adapter implementation. [So, I think the next step is to determine how common this scenario will be] across community maintained adapters.

Of the three custom materializations you mention, `materializedview` will soon become a core materialization in the next Core minor version, `1.6.0`, which is due in less than 10 weeks from today.

So your remaining ask is for contracts on `sources` and `sinks`. Not saying we can't do this work, but I also wanted to ask for an update bc wasn't their some thought to move source/sink config to Terraform and/or not be dbt "models" anymore?

**Reply by morsapaes** [reply]
Thanks for clarifying, @dataders! Being able to support any custom materialization would be ideal, but I understand that not many adapters have this requirement (the Athena folks seem to have found a [workaround](https://github.com/dbt-athena/dbt-athena/pull/284), for example). It's fine to wait until materialized views are in `dbt-core`! I'll add some exceptions to make this clear for the `v1.5.0` bump, in the mean time.

> So your remaining ask is for contracts on sources and sinks. Not saying we can't do this work, but I also wanted to ask for an update bc wasn't their some thought to move source/sink config to Terraform and/or not be dbt "models" anymore?

We'd like to avoid coupling dbt with other tools, since not all `dbt-materialize` users are using Terraform. But we're working on Materialize-native alternatives to manage sources and sinks that will likely allow us to deprecate these materializations in the future — yeah. It'd be cool if the solution was generalizable to any custom materialization, but the priority is definitely `materializedview`!

**Comment by nikitasinghvi** [comment]
Hello, I am exploring dbt-databricks==1.5.1 (it uses dbt-core 1.5.0).
I am trying to implement a contract for a model materialized as a table. I get a compile-time error with dbt run **Model contracts are not currently supported.**

I believe contracts work for materialized tables. Is there something I am missing
```
Compilation Error in model test (models/aac/realtime/test_model_versions/test_version_001.sql)
14:56:12    Model contracts are not currently supported.
14:56:12    
14:56:12    > in macro databricks__persist_constraints (macros/adapters.sql)
14:56:12    > called by macro persist_constraints (macros/adapters.sql)
14:56:12    > called by macro materialization_table_databricks (macros/materializations/table.sql)
14:56:12    > called by model test (models/aac/realtime/test_model_versions/test_version_001.sql)
```

My model files are as follows
test.yml
`version: 2
models:
- name: test
  latest_version: version_002
  config:
    materialized: table
    contract:
      enforced: true
  columns:
    - name: test_column_string
      data_type: string
    - name: test_column_int
      data_type: int
  versions:
    - v: version_001
      defined_in: test_version_001
    - v: version_002
      defined_in: test_version_002
      columns:
        - include: all
          exclude: [ test_column_int ]
`

test_version_001.sql
`select True AS test_column_bool, 1 as test_column_int`


test_version_002.sql
`select  1 as test_column_int`

**Reply by dataders** [reply]
@nikitasinghvi this is better suited as [an issue on dbt-databricks](https://github.com/databricks/dbt-databricks/issues/new?assignees=&labels=bug%2C+triage&projects=&template=bug_report.md&title=). Do you mind opening an issue over there? If you look at [dbt-databricks v1.5.0 release notes](https://github.com/databricks/dbt-databricks/releases/tag/v1.5.0), you'll see that your bug is actually intended behavior.

That said, https://github.com/databricks/dbt-databricks/pull/339 tells me that this support has been merged to the main branch and will likely be soon released as `dbt-databricks v1.5.2`

If you pip install the package directly from GitHub, you might be able to experiment with this new feature

```sh
pip install git+https://github.com/databrickslabs/dbt-databricks.git
```

**Reply by nikitasinghvi** [reply]
Thank you for sharing all the links @dataders, I really appreciate it. I just saw merged but didn't realise it is not for 1.5.1
Let me raise it with databricks.

**Comment by sdebruyn** [comment]
@dataders Could you clarify the logging TBC?

**Comment by genzgd** [comment]
@dataders I know we're a bit late to the party, but ClickHouse has case sensitive table and database (schema) names, so the current adapter `cache` (where all keys are automatically lower-cased) just doesn't work.  Are there any plans to make this more flexible?  At the moment I'm inclined to just disable the cache entirely, since duplicate table names (except for case) I'm guessing would just break things in nasty ways.

**Reply by dbeatty10** [reply]
@genzgd Could you share more what you mean by "ClickHouse  has case sensitive table and database (schema) names"?

Is it that _both_ of these would create a table with a case-sensitive name even though only the latter has a quoted identifier? Or something else?

```sql
CREATE TABLE My_Table ...
CREATE TABLE "My_Table" ...
```

We'd be willing to consider adapter-level interfaces to accommodate tricky edge cases (like the case-sensitivity of table names), but we'd need to better understand what the key database behaviors are.

**Reply by genzgd** [reply]
Thanks for the response -- in ClickHouse quoted identifiers change nothing with respect to table names; both of those attempt to create the same table with a mixed case name.  (Quoted identifiers are only useful for allowing some special characters).  And "My_Table" is a different table than "my_table" or "MY_TABLE", and three tables with all three of those names could exist on the same server (at least on Linux).

Similarly "DB1" and "db1" databases (ClickHouse has no "schema" level) could both exist on the same server, and quoting in the CREATE statement again has no relevant effect.  

We are also having issues with the cache related to the fact that the cache expects both database and schema to be valid strings, but again ClickHouse doesn't actually have schemas, so we reuse the database for that and the relation database property is `None`.  I *think* we can work around that, but in general the fact that the whole cache implementation is hard coded and has no customization or plugin hooks is making it challenging.
